### PR TITLE
Combine packages when package already exists

### PR DESCRIPTION
**Pull request by Siecje**
Fixes https://github.com/takluyver/pynsist/issues/103

Uses https://docs.python.org/3/distutils/apiref.html#distutils.dir_util.copy_tree which will overwrite any existing files.

### CODE REVIEW

File: nsist/pypi.py [hunk 1]
```diff
@@ -181,7 +182,15 @@ def extract_wheel(whl_file, target_dir):
     for p in td.iterdir():
         if p.suffix not in {'.data', '.dist-info'}:
             if p.is_dir():
-                shutil.copytree(str(p), str(target / p.name))
+                try:
+                    shutil.copytree(str(p), str(target / p.name))
+                except OSError as e:
```
Comment (takluyver) [inline nsist/pypi.py]: If the distutils behaviour is what we want in any case, does it make sense to call that directly rather than trying `shutil` first?

Comment (Siecje) [inline nsist/pypi.py]: Yes, I just didn't know if you wanted to change the existing behaviour.

Comment (takluyver) [inline nsist/pypi.py]: I'm not entirely sure either, but your example with PyQtCharts is fairly compelling. And if it's going to change anyway...

It might be worth checking that the packages add different files to the shared directory, and throwing an error if there's a conflict at the file level. But that can be in a separate PR.

File: nsist/pypi.py [hunk 2]
```diff
@@ -181,7 +182,8 @@ def extract_wheel(whl_file, target_dir):
     for p in td.iterdir():
         if p.suffix not in {'.data', '.dist-info'}:
             if p.is_dir():
-                shutil.copytree(str(p), str(target / p.name))
+                # If the dst directory already exists, this will combine them
```
Comment (takluyver) [inline nsist/pypi.py]: Can the comment explicitly say that this is why we use this rather than shutil.copytree? Thanks.

**Comment by Siecje** [comment]
I just notice that you have `merge_dir_to()` which could also be used.

**Comment by takluyver** [comment]
Aha! Yes, I think it would be neat to use `merge_dir_to` here, especially since it already works with Path objects.

**Comment by takluyver** [comment]
You'll need to ensure that the destination directory exists first, though.

**Comment by Siecje** [comment]
I don't know how to use `merge_dir_to` without passing it a `pathlib.Path` object, which was added in `Python 3.4`.

`pynsist` says it supports Python 3.3+

**Comment by takluyver** [comment]
`p` and `target` in that code are already Path objects, so you just use them without calling `str()` on them. We use a backport of pathlib to support older versions of Python.

**Comment by Siecje** [comment]
I'm taking a break from this.

I have another branch that uses `merge_dir-to` https://github.com/Siecje/pynsist/commit/3963c661b2f7547be4051c1292645c4ac33cd5d6

The problem I am running into is the test case with `PyQt5` and `PyQt5Chart` must contain the same file.

```
Traceback (most recent call last):
  File "C:\Users\cody\Anaconda3\lib\runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\cody\Anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\cody\Desktop\qml-testing\venv\Scripts\pynsist.exe\__main__.py", line 9, in <module>
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\__init__.py", line 554, in main
    ).run(makensis=(not options.no_makensis))
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\__init__.py", line 490, in run
    self.prepare_packages()
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\__init__.py", line 372, in prepare_packages
    py_version=self.py_version, bitness=self.py_bitness)
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 212, in fetch_pypi_wheels
    extract_wheel(whl_file, target_dir)
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 195, in extract_wheel
    raise e
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 188, in extract_wheel
    merge_dir_to(p, target.joinpath(p))
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 161, in merge_dir_to
    shutil.copy2(str(p), str(dst_p))
  File "C:\Users\cody\Anaconda3\lib\shutil.py", line 251, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "C:\Users\cody\Anaconda3\lib\shutil.py", line 98, in copyfile
    raise SameFileError("{!r} and {!r} are the same file".format(src, dst))
shutil.SameFileError: 'C:\\Users\\cody\\AppData\\Local\\Temp\\tmpa7eaqg0y\\PyQt5\\pylupdate.pyd' and 'C:\\Users\\cody\\AppData\\Local\\Temp\\tmpa7eaqg0y\\PyQt5\\pylupdate.pyd' are the same file
```

On a unrelated note how do you do `pip install -e` with flit on Windows?

I get an error when I try `flit -f flit.ini install --symlink`

```
Traceback (most recent call last):
  File "C:\Users\cody\Anaconda3\lib\runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\cody\Anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\cody\Desktop\qml-testing\venv\Scripts\flit.exe\__main__.py", line 9, in <module>
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\flit\__init__.py", line 95, in main
    symlink=args.symlink, deps=args.deps).install()
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\flit\install.py", line 333, in install
    self.install_directly()
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\flit\install.py", line 248, in install_directly
    os.symlink(str(self.module.path.resolve()), dst)
OSError: symbolic link privilege not held
```

**Comment by takluyver** [comment]
PyQtChart doesn't appear to contain `pylupdate.pyd`, so I think it must be to do with trying the copy twice. Maybe do a mkdir() before, catching EEXIST, and then calling merge_dir_to once? No rush, though - we all need breaks :-)

I haven't really got a good answer for development installs on Windows. I think there's some way of giving yourself the symbolic link privilege, but I don't know how, nor whether it's a good idea. There's something called NTFS junction points which it may be possible to use. Or maybe it would be best to have a daemon which watches files in packages you're developing and copies them to the installed location. That would ultimately be more flexible because it could also carry out build steps if necessary.

**Comment by Siecje** [comment]
https://github.com/Siecje/pynsist/commit/9766b4147ce3ac4f22409466b16fbd7490c46a4a
I still get the same traceback.

This PR (using distutils.dir_util.copy_tree) does work, if you just want to merge it.

**Comment by takluyver** [comment]
Ah, I know what it is - when you make the destination path with `target.joinpath(p)`, you need `p.name` to get just the last component. You can join path objects using the `/` operator, as the old code does, so you can use `target / p.name` as the destination path.

**Comment by Siecje** [comment]
Each sub directory needs to already exist. https://github.com/Siecje/pynsist/commit/3ff1e97c12b4b9eb514972e54d5c2738b6f69bc0

```shell
(venv) C:\Users\cody\Desktop\qml-testing>pynsist.exe installer.cfg
Unpacking Python...
Copying packages into build directory...
Using cached wheel: PyQt5-5.7.1-5.7.1-cp34.cp35.cp36.cp37-none-win32.whl
Traceback (most recent call last):
  File "C:\Users\cody\Anaconda3\lib\runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\cody\Anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\cody\Desktop\qml-testing\venv\Scripts\pynsist.exe\__main__.py", line 9, in <module>
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\__init__.py", line 554, in main
    ).run(makensis=(not options.no_makensis))
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\__init__.py", line 490, in run
    self.prepare_packages()
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\__init__.py", line 372, in prepare_packages
    py_version=self.py_version, bitness=self.py_bitness)
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 204, in fetch_pypi_wheels
    extract_wheel(whl_file, target_dir)
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 187, in extract_wheel
    merge_dir_to(p, target / p.name)
  File "c:\users\cody\desktop\qml-testing\venv\lib\site-packages\nsist\pypi.py", line 160, in merge_dir_to
    shutil.copy2(str(p), str(dst_p))
  File "C:\Users\cody\Anaconda3\lib\shutil.py", line 251, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "C:\Users\cody\Anaconda3\lib\shutil.py", line 115, in copyfile
    with open(dst, 'wb') as fdst:
FileNotFoundError: [Errno 2] No such file or directory: 'build\\nsis\\pkgs\\PyQt5\\pylupdate.pyd'
```

I guess the next thing to do is write a failing test for `merge_dir_to` and update it.

**Comment by takluyver** [comment]
Merged #105 instead.
